lib/trino-hive-formats/src/main/java/io/trino/hive/formats/line/text/TextLineReaderFactory.java

```diff
@@ -77,11 +77,12 @@ public LineReader createLineReader(
             LineReader lineReader = new TextLineReader(inputStream, fileBufferSize, start, length);
 
             //  Only skip header rows when the split is at the beginning of the file
-            if (headerCount > 0) {
+            if (headerCount > 0 && start == 0) {
                 skipHeader(lineReader, headerCount);
             }
 
-            if (footerCount > 0) {
+            //  Only skip footer rows when the split is at the end of the file
+            if (footerCount > 0 && (start + length == inputFile.length())) {
```

Comment:
When there are footer rows, the file cannot be split, this is because the split could be smaller than the footer section, and there is no way to know that until you read the file.  This is checked when the split is created [here](https://github.com/trinodb/trino/blob/53b993f853ec190e657acb10a7a05a107de28753/plugin/trino-hive/src/main/java/io/trino/plugin/hive/BackgroundHiveSplitLoader.java#L441).  This means that if the case footerCount > 0 and the split does not cover the entire file, and exception should be raised.

Reply:
`inputFile.length()` can sometimes be smaller than the split length in the case of legacy client-side encrypted S3 files, since there is special handling in the legacy S3 file system to fetch the actual size of the file (without the encryption padding).

Note that in `LinePageSourceFactory`, `estimatedFileSize` will be the size from the file listing, which is used for split generation, and thus `estimatedFileSize == size` should be true when the split covers the full file. I think we should change `LinePageSourceFactory` to do `trinoFileSystem.newInputFile(path, estimatedFileSize)` to avoid an remote read and to ensure that the sizes are always the same. (and avoid any weirdness with files being replaced)

Reply:
So with the change to `LinePageSourceFactory`, we can change this check to be
```java
if (footerCount > 0) {
    checkArgument((start == 0) && (length == inputFile.length()), "split must cover entire file for footers");
    ...
}
```

Reply:
> `inputFile.length()` can sometimes be smaller than the split length in the case of legacy client-side encrypted S3 files

I was rather reluctant to use this method to be honest.
It is good that https://github.com/trinodb/trino/pull/18268 avoids using it.